### DISCUSSION TITLE
Nice to have…

### DISCUSSION THREAD

**Discussion by TekyTeky** [Novakasa/brickrail#119] [Ideas]
These are just ideas.. 
The file name in the top of the GUI

Scale of the Edit squares in relation to actual track. 

As it's easy to over run the track when creating it, so it would be good if the red piece of track you select to delete flowed backwards, so you could just keep pressing the delete key to remove track faster.

It took me ages to figure out how to make the block section flow when creating it, you need to make the tiny arrow be in the direction you want... maybe add to the wiki?

Explain about the Settings tab and Render mode and Dynamic/Cached

**Comment by Novakasa** [comment]
Thanks for the input!

> Scale of the Edit squares in relation to actual track.

I'm not sure what you mean by that, could you elaborate?

> As it's easy to over run the track when creating it, so it would be good if the red piece of track you select to delete flowed backwards, so you could just keep pressing the delete key to remove track faster.

Makes sense, but did you know you can delete a whole track section? Sections are what you select before you can create a block.

> It took me ages to figure out how to make the block section flow when creating it, you need to make the tiny arrow be in the direction you want... maybe add to the wiki?

Yeah that used to be different, but it regressed when I refactored some code and I never fixed it.

**Reply by TekyTeky** [reply]
> Thanks for the input!

Welcome.. I'm sure there will be more;-)

> Scale of the Edit squares in relation to actual track.
I'm not sure what you mean by that, could you elaborate?

Does 1 sq equal 2 pieces of track or how many brick studs? Mainly for track planning and scale.. I use an app called BlueBrick ..Lego track planner
https://bluebrick.lswproject.com


> |As it's easy to over run the track when creating it, so it would be good if the red piece of track you select to delete flowed backwards, so you could just keep pressing the delete key to remove track faster.
Makes sense, but did you know you can delete a whole track section? Sections are what you select before you can create a block.

So you can select multiple bits and delete? Also maybe a way to lock the track as sometime you add more in weird locations when relocating the train.

**Reply by Novakasa** [reply]
> Does 1 sq equal 2 pieces of track or how many brick studs? Mainly for track planning and scale.. I use an app called BlueBrick ..Lego track planner https://bluebrick.lswproject.com

Ah I see, you're asking for a to-scale representation of the physical layout. Unfortunately, the way it currently works out it is impossible, since The curve geometry is intended to work nicely with a grid, not to be in scale with LEGO track. For example, if you draw track "diagonally" the lengths are different than doing it vertically or horizontally. With LEGO tracks, diagonal tracks don't really work out that well.
The geometry is inspired by the software "Rocrail" and is more intended as a "schematic" layout. This allows you to be much more flexible with curves and usually saves you a lot of space.

> So you can select multiple bits and delete?

Yup, you can select a track section just like you do when creeating a block.

**Comment by Novakasa** [comment]
implemented the more intuitive tracks section selection (e.g. for blocks) and automatically selecting the next track when deleting one piece now

https://github.com/Novakasa/brickrail/actions/runs/4830487870

**Reply by TekyTeky** [reply]
Thanks for the updates... they make a diff when creating the new layout.

**Reply by TekyTeky** [reply]
Maybe a Batt indicator on the switches too, but maybe poll less frequently?

**Reply by Novakasa** [reply]
There should be a battery indicator for layout_controllers too. But the layout isn't ideal, you might need to incrase the size of the left panel to see it. In current master I changed the UI layout a bit so hopefully it is now easier to see.

**Reply by Novakasa** [reply]
Oh, I just remembered that battery for layout hub currently requires a new firmware, but I didn't distribute that one yet. The train program sources the new communication module dynamically, so that's why the train can already support battery polling (and why the download for train is so slow currently)

**Comment by TekyTeky** [comment]
Battery on the layout hub is good.
I wonder about the space on the left panel for these hubs if people have a lot?
I know you said you plan for other things, so you need the ports and what is connected. I presume it will scroll?
Same for trains I swap mine in and out during a session, should I remove them after I have used it so that I don't have a list of 10 trains?
Although you would lose the settings for that train, I guess you (the user) could "move/drag" the trains being used to the top?
This is what is in one of my BAP sessions. It is ok until you get more then difficult to see the train you are running.
![Screenshot 2023-04-30 150544](https://user-images.githubusercontent.com/87248517/235336641-648579af-b24d-49e1-84b2-eea012aa17a5.png)

**Reply by Novakasa** [reply]
Yeah the hub panel will scroll when many hubs are there.

I am also thinking about only requiring to connect to hubs that are active in the current layout. Maybe you can "disable" trains, so they wouldn't need to be connected when you want to control your layout. Maybe there is something like "train storage" or similar?

**Comment by TekyTeky** [comment]
See on the train panel, maybe you could have Turn next to Stop?
Also, should Start be Cruise... to keep it consistent?
![Screenshot 2023-04-30 154348](https://user-images.githubusercontent.com/87248517/235337739-4d036b2c-2734-4642-9fd7-5cd24357cce4.png)

Just a thought, as I keep doing it with Brickrail: With BAP, people wanted something fast to do the emergency stop rather than find the location in the GUI, so they implemented the space bar as well... so you can run across the room and hit the space bar to stop everthing.

**Reply by Novakasa** [reply]
Agreed with the "start" to "cruise" renaming. Keyboard shortcuts is also planned.

I am also thinking about how to make manual train control a bit more intuitive, but still thinking about how to properly do that.

**Reply by TekyTeky** [reply]
Yes controlling manually is tricky. 
I had some 4D Bricks wifi adapters before they went bust with Covid. I liked their GUI.. it had control for switches like you have, which I think is the best and a wheel, I think, similar to the Lego app. BAP have the slider which is very fiddly. 

To be honest, if I am running manually, I prefer the hand held remotes as you can watch the train and control with your hands without looking.. brain motor cortex functions etc.
Especially if you have multiple trains running.
RocRail is kind of good too.

But in someways I like yours as it is easy.. slow, cruise, fast... stop... turn. With the individual train speeds set that helps a lot.. are they linked? Eg are the manual speeds on the left panel inherited from the train speed setup on the right panel?

But my personal plan was to run this on a seperate loop with switches and randomly run the trains. But the more I use Brickrail the more I like it as a whole of layout controller, especially if I rebuild my layout.

I'd like to get some TrixBrix, but unlike yourself, shipping to Australia isn't cheap. Yours is great with the loops and layers.

**Reply by TekyTeky** [reply]
Oh, and as I said before, having a timetable would be great at some stage😎

**Reply by Novakasa** [reply]
train schedules will be quite some work, so it will take some time.

I guess for now the easiest improvement to manual control would be to have the "switches only" control mode and some keyboard shortcuts for the train control. Incorporating the LEGO remote would be really cool, but would be a lot of work right now and requires rethinking other parts of the communication protocol.

> With the individual train speeds set that helps a lot.. are they linked? Eg are the manual speeds on the left panel inherited from the train speed setup on the right panel?

Yeah, when you manually control the train, it uses the speeds set in the train inspector panel on the right.

**Reply by Novakasa** [reply]
I implemented the "switches only" control devices mode on master branch

**Reply by TekyTeky** [reply]
How do I get/update the master.. I see it but not sure what to do?

**Reply by Novakasa** [reply]
If you go to the "actions" tab, you can see the most recent github builds automatically triggered by new commits:
https://github.com/Novakasa/brickrail/actions
Once you click on a workflow page, stay on the "summary" page to get the artifacts.

**Reply by TekyTeky** [reply]
Ah, I misunderstood what this did.
I thought you just had to connect the layout hub for switches and not the train hubs.. then click the Switch in the GUI to move the Switch.
But this moves the switches when running in Control Devices mode.... but trains don't move... I guess for testing? or have I totally missed something?

**Reply by Novakasa** [reply]
Currently, when you set control devices to "switches", it means that you can still manually control trains in the left panel while being able to control switches by clicking them in the layout.

Currently, it still will only work if all hubs present are connected, even trains). The main point of this mode is that you can manually control trains while switches are able to be controlled as well.

For my layout this will be useful for trains that are on my lower layer and hard to access, but need to be positioned correctly.

**Reply by Novakasa** [reply]
I might need to rethink this whole approach though, as it is probably not intuitive that this is the purpose of this mode.

**Reply by TekyTeky** [reply]
AH, ok, all good.. I understand now... that's still really useful.
It was only about 6:30 am when I was checking.. I have been trying to get up early to see your replies as it is also late for you.

**Reply by Novakasa** [reply]
Haha, please get enough sleep though, Brickrail has been in development for 2 years, so waiting for a reply is allowed to take a few hours :P

**Reply by TekyTeky** [reply]
Ah, but now it is exciting.. I have been waiting 2 years for it to get to this stage!! Serious!

**Comment by TekyTeky** [comment]
3 minor things that would make life easier. 
1. Can the Speed option next to the Random selector be saved please? I always run mine at .8 usually. 
2. In the latest Switch version can there be a Color indicator on the None, Switch & All selector please, like the control devices did before just so it's easy to see if it's on, as I keep forgetting to change it?
3. When you are manually running trains, you don't know which way they are set, Fwd or Rev. So, if you are moving a few you might forget which way one is set to go when you select cruise, and it may hit another train (more of speaking from experience).

**Reply by Novakasa** [reply]
Yeah these make sense. For point 1) what is the reason for you using that setting? Is it the wait times in random targets mode or do you prefer the virtual trains being a bit slower?

If it is the former, I do plan on adding custom wait times anyway, and this setting affecting the real wait times is actually an unintended side effect, but I'll leave it for now.

**Reply by Novakasa** [reply]
BTW, I just added on master a "download" check box for each hub. If you disable it, it will skip downloading the programs, making the setup quite a bit faster.

Only disable it if you have started the program with each of your hubs at least once (this means the hubs have the most recent version of it stored)

For now it is the user's responsibility to judge whether the download is needed, and by default this setting is not remembered, so if you restart brickrail the checkbox will be reset.

In the future I plan to make this system a bit smarter and it will remember whether the last downloaded program is still the most recent one.

**Reply by TekyTeky** [reply]
> Yeah these make sense. For point 1) what is the reason for you using that setting? Is it the wait times in random targets mode or do you prefer the virtual trains being a bit slower?

I like a little time between the next train starting... esp while testing, mainly as I don't like fast trains on a track suspended on wonky tables and the track is not fixed, like this test setup.


> If it is the former, I do plan on adding custom wait times anyway, and this setting affecting the real wait times is actually an unintended side effect, but I'll leave it for now.


That would be great, while my layout is not prototypical, I still like to have things go around, stopping and starting some faster than others etc.. as you saw mine is a very complex layout.

> BTW, I just added on master a "download" check box for each hub. If you disable it, it will skip downloading the programs, making the setup quite a bit faster.

I just downloaded the new master and see the boxes so will try later. It's a good idea as I am impatient like most people, I guess.... and even have the terminal visible to see what's going on.

> Only disable it if you have started the program with each of your hubs at least once (this means the hubs have the most recent version of it stored)
So each time a hub is restarted it clears the code.. but if they are not restarted then the code stays.
So the Connect hubs still works on a restart of the app? 
Is there a timeout if nothing connects after a period of time? eg if I close Brickrail and leave the hubs.

> For now it is the user's responsibility to judge whether the download is needed, and by default this setting is not remembered, so if you restart brickrail the checkbox will be reset.
Yeah cool, good safety net



> In the future I plan to make this system a bit smarter and it will remember whether the last downloaded program is still the most recent one.
Ok, sounds like you have a lot of idea.s.. that's great.

**Reply by Novakasa** [reply]
> So each time a hub is restarted it clears the code.. but if they are not restarted then the code stays.
So the Connect hubs still works on a restart of the app?
Is there a timeout if nothing connects after a period of time? eg if I close Brickrail and leave the hubs.

No, the program is saved even if the hub turns off, but only if it turns off regularly using the button or timeout. If it runs out of battery before being turned off properly, it will forget the program. The hub will also forget the program if you flash the firmware again.

Yes, the hub turns off after a few minutes, but like I said, it should still remember the program.

So technically, as long as you don't flash the firmware again, you can always disable the "download" checkbox.

> mainly as I don't like fast trains on a track suspended on wonky tables and the track is not fixed, like this test setup.

But just to be sure: the virtual train setting doesn't affect the physical train _speed_ in any way, only the wait times.

**Reply by TekyTeky** [reply]
> But just to be sure: the virtual train setting doesn't affect the physical train speed in any way, only the wait times

Yep, got it... that's what I like.

**Reply by Novakasa** [reply]
> In the future I plan to make this system a bit smarter and it will remember whether the last downloaded program is still the most recent one.

This is now implemented on master. However, I did not have chance to test much, so if things seem to be very wrong, you can go back to a previous version.

One caveat to the current approach is that it only remembers whether a hub already should have the program downloaded by the hub name. So if you have a new hub with the same name, Brickrail will not download the program to it. In these cases you should manually activate the "download" checkbox again. Same for when you install new firmware, but use the same hub name as before.

In the future hopefully there will be a good error message if brickrail detects that it couldn't start the program because it was not stored on the hub.

**Reply by TekyTeky** [reply]
OK, will try it today.. thanks.

**Reply by TekyTeky** [reply]
So, do I still need to update your FW onto the hubs as you said it will use the standard Pybricks FW?

**Reply by Novakasa** [reply]
As far as I know, there is no reason to use a specific version of the firmware anymore. I haven't tested the stable version of pybricks in a while, so not sure if I can recommend using just normal pybricks, but there is https://beta.pybricks.com/, which will install the latest beta version of the firmware, which should technically work.

The safest version of the firmware is the one shipped in the brickrail releases, specifically version 2730. It is just a standard build of pybricks master branch. I think in the future I will keep shipping a firmware, just so the users can try exactly the same version as what I am using to test everything, since there might always be some breaking changes in new pybricks firmware.

**Reply by TekyTeky** [reply]
This is how I started up after they had initially downloaded:
Opened layout.
Turned on all hubs.
Select Connect Hubs.
Seems to work fine and is quite fast.
Then I select Control Devices "All" and they are connected immediately.
Saves a heap of time!!

When disconnecting, should I select Disconnect Hubs or is it ok to click the close X?
Some hubs are left with the blue led glowing solid and some slow flash, the slow flashes change to flicker after a few min and then turn off the other 2 stay solid blue.... no big deal. Just thought you should know.
All hubs are running FW 73.

**Reply by Novakasa** [reply]
Normally, it should be fine to close by clicking the "x", since "disconnect hubs" function should still be called.

It is weird that the hubs stayed connected though, (solid light means connected, "breathing" light means the program is still running). Were there any other errors when you were running the trains? As I said I didn't manage to do a lot of testing, so there might still be some bugs.

Maybe try with firmware 2730?

**Reply by Novakasa** [reply]
Oh, and you should _not_ close the BLEServer terminal while hubs are connected

**Reply by TekyTeky** [reply]
I'll try updating those to hubs to 2730.
It has been like that since the beginning, I think.
I didn't have to restart the app or reconnect any hubs... which was great.
Disconnecting is now the slowest part.

I ran individual trains and Random, not the switch option and all went well.
I did use your hidden "feature" of the left mouse button to move the arrow. 
After I did a stop routes some had the arrow in the wrong direction when I continued.

**Reply by Novakasa** [reply]
When you click "disconnect hubs" the hubs will eventually disconnect? Meaning they will not have solid light anymore, but be in "advertising mode" with 2 quick flashes?

**Reply by TekyTeky** [reply]
Yes, Disconnect Hubs and closing the app will disconnect and close the app, just the hubs stay on as I said.

**Reply by TekyTeky** [reply]
Oh, maybe I wasn't clear, they disconnect from the app and it closes, but 2 are still solid, and 2 slow flash, then do the quick 2 flash then turn off, the other 2 are always on. Tomorrow I will reflash to 2730.

**Reply by Novakasa** [reply]
Hmm. So the way it _should_ work is this (and it works like that for me):

Assuming brickrail is running and all hubs are connected and running. If you close Brickrail (press "X" on the title bar, but _not_ for the BLEServer terminal), the brickrail should not immediately close, but the status bar should show status messages like "stopping program" and "disconnecting...". Only after every hub is disconnected (flashing twice with 1 second pause), Brickrail will then actually close. So in total, it will take 5 - 20 seconds to actually close the GUI, depending on how many hubs are connected.

If you press "disconnect hubs", the same process should initiate, only that brickrail does _not_ close after that. I have not tested what happens when you close the GUI while the disconnect process is still ongoing, so for now this is not encouraged.

But in the end, brickrail should never actually allow to be closed if a hub is still showing continuous blue LED (connected) or breathing LED (running program).

If for some reason brickrail is closed while hubs are still running, the hubs will detect after a few seconds that they do not get a response, and stop the program. Then it can take 30-60 seconds for the hub to disconnect by itself. This is however only when things go wrong, usually we should not rely on that.

**Reply by Novakasa** [reply]
The BLEServer terminal should only close by itself automatically after the GUI is closed.

**Reply by TekyTeky** [reply]
Before I updated the Hubs to 2730 I tried again on the disconnecting, and it all worked perfectly.
All like you said and seems even faster, less than 20 secs now, using both methods.
It is weird as I tried many times yesterday and the same 2 hubs always stayed with the led glowing.
This morning it works... I hate these kinds of issues.

Maybe at lunchtime I will see and run the trains for a continuous period.

But starting up now is really great... and on par with my BAP startup.

I am updating the hubs to 2730 anyway.

**Comment by TekyTeky** [comment]
OK, a serious question, are you left-handed?
I ask because it doesn't seem intuitive to move the trains with your right mouse click.
Infact on a laptop with trackpad it is very awkward...  my laptop has a fold over touch screen, and it is not possible to move the trains easily by touch... unless you have a solution.
My plan was to setup the layout and then run the trains via the touchscreen.
I can't really with BAP either, the train sliders are too small.

With 4D Brick I used Linux on small screen 5x7 ins and that worked well.
Also, if you did it on a Mac you could port it to iPad :-)

I always have heaps of ideas for other people's projects!!
Anyway, I really do appreciate all the work you have done!!

**Reply by Novakasa** [reply]
Haha, yes actually I am left handed, however I am still using my mouse with the right hand.

Actually a few months ago the control scheme used to be different. There were 3 different mouse modes "select", "draw track" and "control", and right click was only used for very few things. When I changed to the "edit" and "control" tabs to disallow editing the layout while trains are running, I simplified the control scheme.

Since I have been using desktop PC my whole life and my UX decisions probably come from video games, I didn't really think about these Laptop trackpad issues. I did notice that middle click to drag the view was impossible on trackpads, so I did implement using left click to drag the view.

You are definitely right to highlight these issues, and I will need to think about it. I don't really want to add back different "control modes" again since it introduces some complexity. Would a setting to switch right click/left click already help?

**Reply by TekyTeky** [reply]
Ah.. all makes sense now. Yes, you could have a switch... I was going to go into the settings and change mouse buttons.

Actually, double tap and drag will move the layout, and pinch to zoom will size it. but right click and drag is hard.. but may depend on the laptop .. some don't have buttons anymore either.

**Reply by Novakasa** [reply]
Also I'll look into touch screen stuff, Godot technically is aware of touch screens, but maybe only for android builds...

**Reply by Novakasa** [reply]
It looks like unfortunately touch screen on PC is only implemented in Godot 4, Godot 3 (which brickrail is using) is touch screen aware only on mobile devices...

**Reply by TekyTeky** [reply]
Yeah I just tried some and it was scary ;-)

**Comment by TekyTeky** [comment]
2 really minor things with the inspector….
The drop-down to expose the train speeds always covers them at startup and is hard to see. Even on a large monitor, especially for newbies. The pull bar is not very clear as it is only there when you hover over it. This maybe just as I load so many masters.

The name of the train would be good when you click on it… as the Color dot is small on the virtual layout sometimes, I’m not sure which train I have selected especially if 4 are running. In Run mode.

**Reply by Novakasa** [reply]
> The name of the train would be good when you click on it… as the Color dot is small on the virtual layout sometimes, I’m not sure which train I have selected especially if 4 are running. In Run mode.

I now implemented a "allternative train colors" setting that makes each train color much more obvious.

BTW, it is now also possible to rename trains and blocks by clicking their name in the inspector. This required quite a bit of internal changes, so it might have introduced a number of bugs, and when you load an old layout, train positions might not be the same as before. So after loading an old layout, you might encounter 2 trains at the same block, which of course you should fix in edit more. I am also not sure if the home positions are updated correctly, so it probably makes sense to check those as well. Backup your old layout file before overwriting it with the new brickrail version!

**Comment by TekyTeky** [comment]
Hi, because of my monitor now being remote, I like to have the internal train line Color set to white, this makes it easier to see sensors etc. I changed in the Color in the settings section ok. But now the grid in Edit mode is also white.
If it's an easy change could the grid have its own Color?
If it's too time consuming then please don't worry.
See my pics...
![Layout White tracks - Edit](https://github.com/Novakasa/brickrail/assets/87248517/952f7cd5-3ac0-4746-accb-adc50aa5a9c0)
![Layout White tracks - Run](https://github.com/Novakasa/brickrail/assets/87248517/a70077e7-bf68-4b0f-9396-f3dfb9fac8ae)

**Comment by TekyTeky** [comment]
I know I have asked for this before but would it be possible to have the Space Bar toggle the Emergency Stop, sooner rather than later, please?
The other shortcut keys don't really matter yet, but today when I had to find the mouse and then the button to stop the train, it was really difficult and stressful, especially as a crash is usually imminent.

**Reply by Novakasa** [reply]
I just pushed a commit for this. Build should be ready in a few minutes. Spacebar should now trigger an emergency stop. I have not tested this with actual trains connected yet, so please tell me whether it works.

**Reply by TekyTeky** [reply]
At first test it seems ok, I'll check more today with multiple trains runing... thanks so much !!!

**Reply by TekyTeky** [reply]
I've done quite a bit of testing and it does work well... thanks again.. it saved a few crashes while I am rebuilding the Brickrail layout :)
